### PR TITLE
cuddnnv8 flag Cover std::nullopt case

### PR DESCRIPTION
**Pull request by bhack**
Fixes https://github.com/pytorch/pytorch/pull/134890#pullrequestreview-2419726888

/cc @eqy @Skylion007

### CODE REVIEW

File: aten/src/ATen/native/ConvUtils.h [hunk 1]
```diff
@@ -78,7 +78,11 @@ namespace {
 }
 
 inline bool cudnnv8_enabled_check_debug() {
-  static bool cudnnv8_flag = c10::utils::check_env("TORCH_CUDNN_V8_API_DISABLED") != true;
+  auto cudnnv8_flag_opt = c10::utils::check_env("TORCH_CUDNN_V8_API_DISABLED");
```
Comment (Skylion007) [inline aten/src/ATen/native/ConvUtils.h:81]: Ugh, should it still be static though @cyyever

Comment (cyyever) [reply aten/src/ATen/native/ConvUtils.h:81]: Should keep the static behavior

**Review by cyyever — APPROVED** [review]

**Comment by cyyever** [comment]
@bhack Help grep c10::utils::check_env throughout the codebase and fix similar errors?

**Comment by bhack** [comment]
Yes there are other cases..

**Comment by bhack** [comment]
@cyyever  I suppose that the main issue are `!= true` cases and we have only this one. 
When `== True`  if not set we could consider like if it `False` at least we don't have specific behavior for undef..

Check the list at https://github.com/search?q=repo%3Apytorch%2Fpytorch%20c10%3A%3Autils%3A%3Acheck_env&type=code

**Comment by cyyever** [comment]
@bhack You are right. When comparing against false, nullopt satisfies.

**Review by eqy — COMMENTED** [review]
This was discussed when the env vars were added, the bool check should cover the nullopt case, no?
"truth table":
`TORCH_CUDNN_V8_API_DISABLED` not set, `nullopt, != true -> true`, cuDNN v8 is enabled, correct 
`TORCH_CUDNN_V8_API_DISABLED` set 0, `0, != true -> true`, cuDNN v8 is enabled, correct
`TORCH_CUDNN_V8_API_DISABLED` set 1, `1 != true -> false`. cuDNN v8 is disabled, correct

**Comment by bhack** [comment]
@eqy this seems ok but then we need to recheck https://github.com/pytorch/pytorch/pull/134890#pullrequestreview-2419726888

Can you verify that this simulation is correct?
```c++
#include <iostream>
#include <optional>

std::optional<bool> getValue(bool returnNullopt) {
    if (returnNullopt) {
        return std::nullopt;
    }
    return 1;
}

int main() {
    static bool value = getValue(true) != true;
    static bool flag= (value == 1);
    if (flag) {
        std::cout << "Value is nullopt (not set)." << std::endl;
    } else {
        std::cout << "Value is set to: " << value << std::endl;
    }
    if (!(false && flag)) {
        std::cout << "Warning" << std::endl;
    }
    return 0;
}

```

**Comment by eqy** [comment]
I modified your code snippet to cover all of the cases and I still do not see where the behavior is incorrect. We want to warn when either cuDNN is older than 9.3 OR when the V8+ API is disabled, and another way to write this is we warn unless both conditions are satisfied.
```
int main() {
    static bool value = getValue(true) != true;
    static bool flag= (value == 1);
    if (flag) {
        std::cout << "Value is nullopt (not set)." << std::endl;
    } else {
        std::cout << "Value is set to: " << value << std::endl;
    }
    // cuDNN older than 9.3 and V8 is enabled
    if (!(false && flag)) {
        std::cout << "Warning when cuDNN older than 9.3 and V8 API enabled (correct)" << std::endl;
    }
    // cuDNN newer than 9.3 and V8 is enabled
    if (!(true && flag)) {
        std::cout << "Warning when cuDNN newer than 9.3 and V8 API enabled (incorrect)" << std::endl;
    }
    flag = false;
    // cuDNN older than 9.3 and V8 is disabled
    if (!(false && flag)) {
        std::cout << "Warning when cuDNN older than 9.3 and V8 API disabled (correct)" << std::endl;
    }
    // cuDNN newer than 9.3 and V8 is disabled
    if (!(true && flag)) {
        std::cout << "Warning when cuDNN newer than 9.3 and V8 API disabled (correct)" << std::endl;
    }


    return 0;
}
```
Value is nullopt (not set).
Warning when cuDNN older than 9.3 and V8 API enabled (correct)
Warning when cuDNN older than 9.3 and V8 API disabled (correct)
Warning when cuDNN newer than 9.3 and V8 API disabled (correct)
```

**Comment by bhack** [comment]
So do we need both 9.3 and V8 API enabled to not have the warning?

Cause then here we need `and` here

>  Consider upgrading cuDNN and/or enabling the V8 API for better efficiency.

**Comment by bhack** [comment]
Cause 
>   if (!(false && flag)) {

We have
```c++
        TORCH_WARN_ONCE("cuDNN cannot be used for large non-batch-splittable convolutions"
                        " if the V8 API is not enabled or before cuDNN version 9.3+."
                        " Consider upgrading cuDNN and/or enabling the V8 API for better efficiency.");
        return false;
```

**Comment by eqy** [comment]
I am interpreting this
" if the V8 API is not enabled or before cuDNN version 9.3+."
as being a superset of
"Warning when cuDNN older than 9.3 and V8 API enabled (correct)", more explicitly
 (not enabled is false) OR (cuDNN is older than 9.3 is true) -> true cuDNN _cannot_ be used, so we surface the warning
which is correct. The prose comment is _or_, so it means the warning fires when either condition is present.

**Comment by bhack** [comment]
Just to recap if I have only the `V8 API` enabled do I need to have the warning or not?

**Comment by bhack** [comment]
To simplify currently the case seems
```c++
#include <iostream>
#include <optional>

std::optional<bool> getValue(bool returnNullopt) {
    if (returnNullopt) {
        return std::nullopt;
    }
    return 1;
}

int main() {
    static bool value = getValue(true) != true;
    static bool flag= (value == 1);
    if (flag) {
        std::cout << "Value is nullopt (not set)." << std::endl;
    } else {
        std::cout << "Value is set to: " << value << std::endl;
    }
    if (!(false && flag)) {
        std::cout << "cuDNN cannot be used for large non-batch-splittable convolutions"
                        " if the V8 API is not enabled or before cuDNN version 9.3+."
                        " Consider upgrading cuDNN and/or enabling the V8 API for better efficiency." << std::endl;
    }
    return 0;
}
```

**Comment by bhack** [comment]
So when `cuDNN older than 9.3` and  `TORCH_CUDNN_V8_API_DISABLED` is not set we have 
```c++
std::cout << "cuDNN cannot be used for large non-batch-splittable convolutions"
                        " if the V8 API is not enabled or before cuDNN version 9.3+."
                        " Consider upgrading cuDNN and/or enabling the V8 API for better efficiency." << std::endl;
```

Is that what we want? Logically it seems not.

**Comment by eqy** [comment]
I would say that is what we want as 64-bit would not be enabled and the user should upgrade to enable 64-bit non-batch-splittable convs through cuDNN.

**Comment by bhack** [comment]
So IMHO warning  about the action to `V8 API` enabling is confusing if it is mandatory to have `9.3+`

**Comment by eqy** [comment]
V8 API disablement/enablement is orthogonal to the cuDNN version in this case. cuDNN v9 can still be run with the V8 API disabled.

**Comment by bhack** [comment]
Yes I understand but the message is confusing cause it seems still possible to use it only with V8 API enabled (without 9.3+)

**Comment by bhack** [comment]
Probably as we know the version in the condition we could different the warning right?

**Comment by eqy** [comment]
I think we are disagreeing on the meaning of or: in prose "You cannot drive if you are drunk or do not have a license" is what I mean, not "You cannot drive if you are drunk and do not have a license."

This message will also be mostly irrelevant once the upstream wheels move to 9.5.1: https://github.com/pytorch/pytorch/pull/137978

**Comment by bhack** [comment]
Yes I saw  https://github.com/pytorch/pytorch/pull/137978 

In any case `Consider upgrading cuDNN and/or enabling the V8 API for better efficiency` it is still quite confusing specially cause we know what cuDNN version we have so we could give to the user a better action item.

But in any case let's wait for the 9.5.1 landing on main.

**Pull request closed without merging**